dm/worker/log.go

```diff
@@ -0,0 +1,498 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package worker
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"reflect"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/dm/dm/pb"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/errors"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+// ErrInValidHandler indicates we meet an invaild  Putter/Getter/Deleter
+var ErrInValidHandler = errors.New("handler is nil, please pass a leveldb.DB or leveldb.Transaction")
+
+// Putter is interface which has Put method
+type Putter interface {
+	Put(key, value []byte, opts *opt.WriteOptions) error
+}
+
+// Deleter is interface which has Delete method
+type Deleter interface {
+	Delete(key []byte, wo *opt.WriteOptions) error
+}
+
+// Getter is interface which has Get method
+type Getter interface {
+	Get(key []byte, ro *opt.ReadOptions) ([]byte, error)
+}
+
+// HandledPointerKey is key of HandledPointer which point to the last handled log
+var HandledPointerKey = []byte("!DM!handledPointer")
+
+// Pointer is a logic pointer that point to a location of log
+type Pointer struct {
+	Location int64
+}
+
+// MarshalBinary never return not nil err now
+func (p *Pointer) MarshalBinary() []byte {
```

Comment:
not implement `BinaryMarshaler.MarshalBinary() (data []byte, err error)` (because it never returns a not nil err), but use it directly?

Reply:
`BinaryMarshaler.MarshalBinary() (data []byte, err error)` is better